# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/power/WakeLockJS.h
# Commit: 82e03a404c2f
# Full Hash: 82e03a404c2fe76b1240dce4fa3d0b6bde909a87
# Author: Vincent Hilla <vhilla@mozilla.com>
# Date: 2023-12-05 21:29:44
# Regressor Bug: 1589554
# File Overlap Count: 1
# Description:
#   Bug 1589554 - Part 2: Implement Screen Wake Lock API. r=dom-core,edgar
#   
#   Depends on D189508
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D189509
# ==============================================================================

diff -r b55991835aec -r 82e03a404c2f dom/power/WakeLockJS.h
--- a/dom/power/WakeLockJS.h	Tue Dec 05 15:37:05 2023 +0000
+++ b/dom/power/WakeLockJS.h	Tue Dec 05 15:37:05 2023 +0000
@@ -8,6 +8,7 @@
 #define DOM_WAKELOCKJS_H_
 
 #include "js/TypeDecls.h"
+#include "mozilla/Attributes.h"
 #include "mozilla/dom/WakeLockBinding.h"
 #include "nsIDOMEventListener.h"
 #include "nsIDocumentActivity.h"
@@ -26,6 +27,15 @@
 
 namespace mozilla::dom {
 
+/**
+ * Management class for wake locks held from client scripts.
+ * Instances of this class have two purposes:
+ * - Implement navigator.wakeLock.request which creates a WakeLockSentinel
+ * - Listen for state changes that require all WakeLockSentinel to be released
+ * The WakeLockSentinel objects are held in document.mActiveLocks.
+ *
+ * https://www.w3.org/TR/screen-wake-lock/#the-wakelock-interface
+ */
 class WakeLockJS final : public nsIDOMEventListener,
                          public nsWrapperCache,
                          public nsIDocumentActivity {
@@ -41,7 +51,7 @@
   explicit WakeLockJS(nsPIDOMWindowInner* aWindow);
 
  protected:
-  ~WakeLockJS() = default;
+  ~WakeLockJS();
 
  public:
   nsISupports* GetParentObject() const;
@@ -49,12 +59,37 @@
   JSObject* WrapObject(JSContext* aCx,
                        JS::Handle<JSObject*> aGivenProto) override;
 
-  already_AddRefed<Promise> Request(WakeLockType aType);
+  already_AddRefed<Promise> Request(WakeLockType aType, ErrorResult& aRv);
 
  private:
+  enum class RequestError {
+    Success,
+    DocInactive,
+    DocHidden,
+    PolicyDisallowed,
+    PrefDisabled,
+    InternalFailure
+  };
+
+  static nsLiteralCString GetRequestErrorMessage(RequestError aRv);
+
+  static RequestError WakeLockAllowedForDocument(Document* aDoc);
+
+  void AttachListeners();
+  void DetachListeners();
+
+  Result<already_AddRefed<WakeLockSentinel>, RequestError> Obtain(
+      WakeLockType aType);
+
+  void UnlockAll(WakeLockType aType);
+
   RefPtr<nsPIDOMWindowInner> mWindow;
 };
 
+MOZ_CAN_RUN_SCRIPT
+void ReleaseWakeLock(Document* aDoc, WakeLockSentinel* aLock,
+                     WakeLockType aType);
+
 }  // namespace mozilla::dom
 
 #endif  // DOM_WAKELOCKJS_H_